ledger/store/src/program/finalize.rs

```diff
@@ -250,207 +231,107 @@ pub trait FinalizeStorage<N: Network>: 'static + Clone + Send + Sync {
+            // Remove the existing key-value entries.
+            self.key_value_id_map().remove_map(&(program_id, mapping_name))?;
 
-            // Insert the new key-value ID map.
-            self.key_value_id_map().insert(mapping_id, next_key_value_ids)?;
-            // Insert the new keys and values.
-            for (key_id, key, value) in next_keys_values {
-                self.key_map().insert(key_id, key)?;
-                self.value_map().insert(key_id, value)?;
+            // Insert the new key-value entries.
+            for (key, value) in entries {
+                // Insert the key-value ID.
+                self.key_value_id_map().insert((program_id, mapping_name), key, value)?;
             }
 
             Ok(())
         })?;
 
         // Return the finalize operation.
-        Ok(FinalizeOperation::ReplaceMapping(mapping_id))
+        Ok(FinalizeOperation::ReplaceMapping(to_mapping_id(&program_id, &mapping_name)?))
     }
 
     /// Removes the mapping for the given `program ID` and `mapping name` from storage,
     /// along with all associated key-value pairs in storage.
-    fn remove_mapping(&self, program_id: &ProgramID<N>, mapping_name: &Identifier<N>) -> Result<FinalizeOperation<N>> {
-        // Retrieve the mapping ID.
-        let Some(mapping_id) = self.get_mapping_id_speculative(program_id, mapping_name)? else {
-            bail!("Illegal operation: mapping '{mapping_name}' is not initialized - cannot remove mapping.")
-        };
-        // Retrieve the key-value IDs for the mapping ID.
-        let Some(key_value_ids) = self.key_value_id_map().get_speculative(&mapping_id)? else {
-            bail!("Illegal operation: mapping ID '{mapping_id}' is not initialized - cannot remove mapping.")
-        };
+    fn remove_mapping(&self, program_id: ProgramID<N>, mapping_name: Identifier<N>) -> Result<FinalizeOperation<N>> {
+        // Ensure the mapping name exists.
+        if !self.key_value_id_map().contains_map_speculative(&(program_id, mapping_name))? {
+            bail!("Illegal operation: '{program_id}/{mapping_name}' is not initialized - cannot remove mapping.")
+        }
 
         // Retrieve the mapping names.
-        let mut mapping_names = match self.program_id_map().get_speculative(program_id)? {
+        let mut mapping_names = match self.program_id_map().get_speculative(&program_id)? {
             Some(mapping_names) => cow_to_cloned!(mapping_names),
             None => bail!("Illegal operation: program ID '{program_id}' is not initialized - cannot remove mapping."),
         };
         // Ensure the mapping name exists.
-        if !mapping_names.contains(mapping_name) {
+        if !mapping_names.contains(&mapping_name) {
             bail!("Illegal operation: mapping '{mapping_name}' does not exist in storage - cannot remove mapping.");
         }
         // Remove the mapping name.
-        mapping_names.remove(mapping_name);
+        mapping_names.remove(&mapping_name);
```

Comment:
this can be done faster with the following:
```
        if !mapping_names.remove(&mapping_name) {
            bail!("Illegal operation: mapping '{mapping_name}' does not exist in storage - cannot remove mapping.");
        }
```
if the mapping was not there, nothing changes, and if it was, it gets removed as expected

Reply:
changed in 4a916455b